From 8e04c0735d8049a628bc47ba0a92b8c7bde0b922 Mon Sep 17 00:00:00 2001
From: Pavel Tikhomirov <ptikhomirov@virtuozzo.com>
Date: Wed, 12 Jan 2022 13:34:39 +0300
Subject: [PATCH 166/245] mount: rework skipping external mounts in
 dump_one_mountpoint

Function dump_one_fs already has mnt_is_external_bind check inside, so
there is no point to check pm->external one more time.

Function check_bindmount is intended to check devpts bindmount's master
was opened in right mount namespace, but if bindmount is external mount
there is no point to check this. Let's also skip check for bindmounts of
external mounts.

Signed-off-by: Pavel Tikhomirov <ptikhomirov@virtuozzo.com>
---
 criu/mount.c | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/criu/mount.c b/criu/mount.c
index ef50d0fa2..844f52e4d 100644
--- a/criu/mount.c
+++ b/criu/mount.c
@@ -1674,13 +1674,12 @@ static int dump_one_mountpoint(struct mount_info *pm, struct cr_img *img)
 	if (me.fstype == FSTYPE__AUTO)
 		me.fsname = pm->fsname;
 
-	if (!pm->external) {
-		if (!pm->dumped && dump_one_fs(pm))
-			return -1;
+	if (!pm->dumped && dump_one_fs(pm))
+		return -1;
 
-		if (!fsroot_mounted(pm) && pm->fstype->check_bindmount && pm->fstype->check_bindmount(pm))
-			return -1;
-	}
+	if (!mnt_is_external_bind(pm) && !fsroot_mounted(pm) && pm->fstype->check_bindmount &&
+	    pm->fstype->check_bindmount(pm))
+		return -1;
 
 	if (pm->mnt_id == CRTIME_MNT_ID) {
 		pr_info("Skip dumping cr-time mountpoint: %s\n", pm->mountpoint);
-- 
2.35.1

